// firebase-admin/lib/app/firebase-namespace.js
/*! firebase-admin v13.4.0 */
"use strict";
/*!
 * @license
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultNamespace = exports.FirebaseNamespace = exports.FirebaseNamespaceInternals = void 0;
const lifecycle_1 = require("./lifecycle");
const credential_factory_1 = require("./credential-factory");
const index_1 = require("../utils/index");
/**
 * Internals of a FirebaseNamespace instance.
 */
class FirebaseNamespaceInternals {
    constructor(appStore) {
        this.appStore = appStore;
    }
    /**
     * Initializes the App instance.
     *
     * @param options - Optional options for the App instance. If none present will try to initialize
     *   from the FIREBASE_CONFIG environment variable. If the environment variable contains a string
     *   that starts with '{' it will be parsed as JSON, otherwise it will be assumed to be pointing
     *   to a file.
     * @param appName - Optional name of the FirebaseApp instance.
     *
     * @returns A new App instance.
     */
    initializeApp(options, appName) {
        const app = this.appStore.initializeApp(options, appName);
        return extendApp(app);
    }
    /**
     * Returns the App instance with the provided name (or the default App instance
     * if no name is provided).
     *
     * @param appName - Optional name of the FirebaseApp instance to return.
     * @returns The App instance which has the provided name.
     */
    app(appName) {
        const app = this.appStore.getApp(appName);
        return extendApp(app);
    }
    /*
     * Returns an array of all the non-deleted App instances.
     */
    get apps() {
        return this.appStore.getApps().map((app) => extendApp(app));
    }
}
exports.FirebaseNamespaceInternals = FirebaseNamespaceInternals;
const firebaseCredential = {
    cert: credential_factory_1.cert, refreshToken: credential_factory_1.refreshToken, applicationDefault: credential_factory_1.applicationDefault
};
/**
 * Global Firebase context object.
 */
class FirebaseNamespace {
    /* tslint:enable */
    constructor(appStore) {
        // Hack to prevent Babel from modifying the object returned as the default admin namespace.
        /* tslint:disable:variable-name */
        this.__esModule = true;
        /* tslint:enable:variable-name */
        this.credential = firebaseCredential;
        this.SDK_VERSION = (0, index_1.getSdkVersion)();
        /* tslint:disable */
        // TODO(jwenger): Database is the only consumer of firebase.Promise. We should update it to use
        // use the native Promise and then remove this.
        this.Promise = Promise;
        this.INTERNAL = new FirebaseNamespaceInternals(appStore ?? new lifecycle_1.AppStore());
    }
    /**
     * Gets the `Auth` service namespace. The returned namespace can be used to get the
     * `Auth` service for the default app or an explicitly specified app.
     */
    get auth() {
        const fn = (app) => {
            return this.ensureApp(app).auth();
        };
        const auth = require('../auth/auth').Auth;
        return Object.assign(fn, { Auth: auth });
    }
    /**
     * Gets the `Database` service namespace. The returned namespace can be used to get the
     * `Database` service for the default app or an explicitly specified app.
     */
    get database() {
        const fn = (app) => {
            return this.ensureApp(app).database();
        };
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        return Object.assign(fn, require('@firebase/database-compat/standalone'));
    }
    /**
     * Gets the `Messaging` service namespace. The returned namespace can be used to get the
     * `Messaging` service for the default app or an explicitly specified app.
     */
    get messaging() {
        const fn = (app) => {
            return this.ensureApp(app).messaging();
        };
        const messaging = require('../messaging/messaging').Messaging;
        return Object.assign(fn, { Messaging: messaging });
    }
    /**
     * Gets the `Storage` service namespace. The returned namespace can be used to get the
     * `Storage` service for the default app or an explicitly specified app.
     */
    get storage() {
        const fn = (app) => {
            return this.ensureApp(app).storage();
        };
        const storage = require('../storage/storage').Storage;
        return Object.assign(fn, { Storage: storage });
    }
    /**
     * Gets the `Firestore` service namespace. The returned namespace can be used to get the
     * `Firestore` service for the default app or an explicitly specified app.
     */
    get firestore() {
        let fn = (app) => {
            return this.ensureApp(app).firestore();
        };
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const firestore = require('@google-cloud/firestore');
        fn = Object.assign(fn, firestore.Firestore);
        // `v1beta1` and `v1` are lazy-loaded in the Firestore SDK. We use the same trick here
        // to avoid triggering this lazy-loading upon initialization.
        Object.defineProperty(fn, 'v1beta1', {
            get: () => {
                return firestore.v1beta1;
            },
        });
        Object.defineProperty(fn, 'v1', {
            get: () => {
                return firestore.v1;
            },
        });
        return fn;
    }
    /**
     * Gets the `MachineLearning` service namespace. The returned namespace can be
     * used to get the `MachineLearning` service for the default app or an
     * explicityly specified app.
     */
    get machineLearning() {
        const fn = (app) => {
            return this.ensureApp(app).machineLearning();
        };
        const machineLearning = require('../machine-learning/machine-learning').MachineLearning;
        return Object.assign(fn, { MachineLearning: machineLearning });
    }
    /**
     * Gets the `Installations` service namespace. The returned namespace can be used to get the
     * `Installations` service for the default app or an explicitly specified app.
     */
    get installations() {
        const fn = (app) => {
            return this.ensureApp(app).installations();
        };
        const installations = require('../installations/installations').Installations;
        return Object.assign(fn, { Installations: installations });
    }
    /**
     * Gets the `InstanceId` service namespace. The returned namespace can be used to get the
     * `Instance` service for the default app or an explicitly specified app.
     */
    get instanceId() {
        const fn = (app) => {
            return this.ensureApp(app).instanceId();
        };
        const instanceId = require('../instance-id/instance-id').InstanceId;
        return Object.assign(fn, { InstanceId: instanceId });
    }
    /**
     * Gets the `ProjectManagement` service namespace. The returned namespace can be used to get the
     * `ProjectManagement` service for the default app or an explicitly specified app.
     */
    get projectManagement() {
        const fn = (app) => {
            return this.ensureApp(app).projectManagement();
        };
        const projectManagement = require('../project-management/project-management').ProjectManagement;
        return Object.assign(fn, { ProjectManagement: projectManagement });
    }
    /**
     * Gets the `SecurityRules` service namespace. The returned namespace can be used to get the
     * `SecurityRules` service for the default app or an explicitly specified app.
     */
    get securityRules() {
        const fn = (app) => {
 !   3 k3  !   (e e  !   5 5 n!   5 5 !   p= q= @!   q= r= (!   r= r= P!   s= t=   !   r= s= `!   5 5 !   e f !   / h/ g!    n0 o0 8l!   0? ? !   ? ,? !   P5 5 X !    e Pe p!    h i  !   lm0 m0 !   |/ }/ h !   @? ? \ !   _4 7`4 !   8`4 b4 8!   0]4 _4 !   ?  ?  !   @? Ƈ? ć!   Ї? ? 4!   ? @? !    = _=  !   = r= $!   =  = h !   \= q=  !   |0 }0 X !   pv= ;w= !   <w= w= !   u= nv=  !   p36 36 ع !   u= ^u= X !   `u= u= t!   3A 05A 4 !   A A !   H@ @ X !   dA eA X !   EA WFA \ !   l5A 5A ع !   \HA HA ` !   `= = <!   @ @  !   `nU oU !   .A ~/A  !   x1A A2A |!   > >  !   "6 =#6  !   0A u1A \!   `G RG -!   >  > p!   HG UG -!   G G -!   n ?n  !   o ;p 8 !   pd= e= X!   @t Vt  !   `t t  !   t Su |!   @6 X6 !   v -v X !   Pv v !   X6 ־6 ( !   v 0w !   `y \{ !   \{ | !   | | 	!   } ~ 	!   ,  8	!      L	!       !      t	!    u 	!   x ]  !   6 )6  !   `6 6  !   6 m6  !   )7 )7 X !   )7 *7  !    k 	!   7 7 D!   l  	!    S X !   T  	!      !     X !     [ @
!   X  
!    ߢ 
!     
!    R 
!     
!    <  !   <  h !    Ҥ  !     
!    4  !   @  !    ,  !   ,   !     8!    ) !   , l !   l   !     !    [ !   \  (!   d  D!    <  !     x!   @  !     !    2 ع !      !   о п !     X !   ( 9 $
!   t' ' &!    [ \
!   \   !      !   p@ n@  !    ' 
!   D$ !$  !   0 [ ع !   ؾ6 ?6 ț!   ]B ^B '!   8D D  !   \  
!   $B $B 4!   @6 6 !   ^B ^B ع !     x !   $A KA X !    @ `@ D!   LA A P!   GB HB Pv!   ^B D_B  !   D_B e_B ع !    $ $  !   $ $ ع !   A A  !   A IA  !    T 
!   @B B  !   B $B !   LA A !   LK K 0M!   @  ع !    u@ u@  !    3I i5I 8!     @!   . \. _!   \. . !   @ }  !      !     X !   `@ #@  !   mI &oI :!   (oI $pI @:!   dJ J @J!    - h!   0 $ !     !    ,  !   K K TM!   K K =!   I I X !   J ЈJ hH!     !    C 0!   D # T!   $  x!      !   r+ Os+ _!   I  I <!    I I x!   d2J %3J  !   K K g!   K TK  !   `K K !   I I ع !   I \I  !   0  !    J J J!     !    n  !   f ;f  !   z, {, ( !   ;* ;*  !    2 !   2 3 4 4=vp »p 0!   	 	  !    o, 5p, L!   p 2p p!    p 2p p!   Ke Le !   , , H!   	 	  !    	 N	  !   P	 	  !   P, ,  !   d, 
j, <L!   Le Me !    Ne -Pe !   	 	 !   D/ c/  !   	 ~	  !   	 K	 h !   le/ Mh/ t !   D/ /  g!   	 	 !   %4 S%4 ع !   `%4 n&4 ` !   p/ J/ Ԧ!   	 	  !   8l4 o4 !   _* d* T;!   d* Oe*  !   	 !	  !    !	 !	 8 !   @#	 %	 !   &	 6'	  !   @'	 '	 \ !   '	 (	 8 !   (	 3)	 ع !   
1 
1  !   K* L* ;!   <h4 j4  !   $1	 {2	 !   3	 4	 !    f f X !   4	 4	 X !   p7	 q8	 !   8	 8	 X !   \9	 9	 ع !   T:	 <	 !   ; Ԍ;  !   @	 -B	 D!   0B	 B	 ع !   tD	 E	 l!   F	 lG	 !   lG	 I	 !   I	 MK	 !   PK	 xK	 <!   L	 /M	  !   N	 _N	  !   `N	 3O	 @!   T	 U	 \ !   <5 @=5  !    5 5 !   ,5 5  !   5 x5 0!   |5 5  !   5 ɫ5 !   آ4 b4 !   p5 :5 P!   5 5 x!   5 e5 8 !   @5 y5 !   $V	 eV	  !   W	 W	 h !   4 Z4 P!   W	 X	 h !   55 75 X!   \4 N4 l!   5 5  !   ?5 0C5 \!   + + TH!   d  4 =p ip ,!    b	 fb	 D !     T p !p T!   d	 e	  !   e	 Tf	 t!   ;* %<* :!   ]* ]* X !   ( =( X !   ( A(  !   f kf #!   @5 5  !   pf f"f #!   
